Sources/OSLogClient/Public/LogClient.swift

```diff
@@ -0,0 +1,229 @@
+        if await isPollingEnabled, await shouldPauseIfNoRegisteredDrivers, await logDriverRegistry.drivers.isEmpty {
+            softStopPolling()
+        }
+    }
+
+    /// Will force an immediate poll of logs on a detached task.
+    /// **Note:** This does not reset or otherwise alter the current interval driven polling.
+    /// - Parameter date: Optional date to query from. Leave `nil` to query from the last time logs were polled (default behaviour).
+    public func forcePoll(from date: Date? = nil) {
+        // Generate task
+        let taskId: UUID = .init()
+        let pollTask: Task<(), Error> = Task(priority: .userInitiated) { [weak self, taskId] in
+            guard let self else { return }
+            await self.pollLatestLogs(from: date)
+            self.immediatePollTaskMap.removeValue(forKey: taskId)
+        }
+        immediatePollTaskMap[taskId] = pollTask
+    }
+    
+    /// Will assign the given Bool flag to the ``LogClient/shouldPauseIfNoRegisteredDrivers`` property.
+    ///
+    /// - Note: When `true` if all drivers are deregistered, polling will cease until valid drivers are registered again.
+    /// - Parameter flag: Bool whether enabled.
+    public func setShouldPauseIfNoRegisteredDrivers(_ flag: Bool) async {
+        await config.setShouldPauseIfNoRegisteredDrivers(flag)
+    }
+
+    // MARK: - Helpers: Polling
+    
+    /// Will cancel any pending poll tasks, but not assign the `isPollingEnabled` flag to false.
+    /// This is called when all drivers are deregistered but the consumer has not explicitly stopped polling.
+    func softStopPolling() {
+        pendingPollTask?.cancel()
+        pendingPollTask = nil
+    }
+
+    /// Will assess the given entry and ensure it is both a valid `OSLogEntryLog` instance.
+    /// - Parameter entry: The entry to assess.
+    /// - Returns: `OSLogEntryLog` or `nil`
+    func validateLogEntry(_ entry: OSLogEntry) -> OSLogEntryLog? {
+        entry as? OSLogEntryLog
+    }
+
+    /// Will poll logs since the last processed time position and send to any registered drivers for validation and processing.
+    /// - Parameter date: Optional date to request logs from. Leave this `nil` to default to the `lastProcessed` property.
+    func pollLatestLogs(from date: Date? = nil) async {
+        do {
+            var predicate: NSPredicate?
+            let lastProcessed = await config.lastProcessed
+            if let fromDate = date ?? lastProcessed {
+                predicate = datePredicate.withSubstitutionVariables(["DATE": fromDate])
+            }
+            let items = try logStore.getEntries(matching: predicate).compactMap(validateLogEntry)
+            let sortedItems = items.sorted(by: { $0.date <= $1.date })
+            // Broadcast logs to drivers
+            await sortedItems.asyncForEach { log in
+                await logDriverRegistry.drivers.asyncForEach {
+                    $0.processLogIfValid(log)
+                }
+            }
```

Comment:
yeah I get that, so you thinking along lines of (contrived).

```swift
// LogDriver.swift
    func isValidLog(_ log: OSLogEntryLog) -> Bool {
        // existing validation logic
    }

    func processLogs(_ logs: [OSLogEntryLog]) {
        let valids = logs.filter(isValidLog)
        valids.forEach {
            let logLevel = LogLevel($0.level)
            #if os(macOS)
            // existing process invocation
            #else
            // existing process invocation
            #endif
        }
    }

// LogClient.swift:

func pollLatestLogs(from date: Date? = nil) async {
        do {
            ...
            let items = try logStore.getEntries(matching: predicate).compactMap(validateLogEntry)
            let drivers = await logDriverRegistry.drivers
            drivers.forEach {
                $0.processLogs(items)
            }
            ...
}
```

The sorting (from my notes) was performed as I needed to get the most recent date/timestamp to store to the `lastProcessed` property - so would need to do that still - could leave where it was or do afterwards etc?

Reply:
Yeah basically that. for loop over forEach, just to minimise performance impact, albeit minimal.

If you leave the processing and sorting to the drivers to perform async then you could avoid sorting in the client and push that performance decision to the driver, then for the last processed date it would be the max function on all the logs.

Alternatively last processed _could_ be invocation time, since fetching logs is from the last invocation up to now. Could even tweak the predicate to include a range to avoid potentially missing a log.

Reply:
Latest commit has these changes